File: senam_dashboard/src/lib/media-url.ts
const API_BASE = (import.meta.env.VITE_API_BASE_URL as string).replace(
  /\/+$/,
  '',
);
const UPLOADS_RAW_PREFIX = '/v1/uploads/raw/';

const UPLOAD_PURPOSES = [
  'profile_photo',
  'kyc_document',
  'review_photo',
  'company_logo',
  'company_cover',
  'portfolio_photo',
  'gallery_photo',
  'banner_image',
  'category_icon',
  'service_icon',
  'service_image',
] as const;

function isUploadObjectKey(value: string): boolean {
  return UPLOAD_PURPOSES.some((p) => value.startsWith(`${p}/`));
}

/** Host-independent path for DB / form state. */
export function toStoredMediaPath(value: string): string {
  const trimmed = value.trim();
  if (!trimmed) return trimmed;

  try {
    const u = new URL(trimmed);
    const idx = u.pathname.indexOf(UPLOADS_RAW_PREFIX);
    if (idx >= 0) return u.pathname.slice(idx);
    return trimmed;
  } catch {
    /* relative path or bare key */
  }

  if (trimmed.startsWith(UPLOADS_RAW_PREFIX)) return trimmed;

  if (isUploadObjectKey(trimmed)) {
    return `${UPLOADS_RAW_PREFIX}${trimmed}`;
  }

  return trimmed;
}

/** Absolute URL for <img src>. Rewrites legacy upload hosts to VITE_API_BASE_URL. */
export function resolveMediaUrl(value: string | null | undefined): string {
  if (!value) return '';
  const stored = toStoredMediaPath(value);
  if (stored.startsWith(UPLOADS_RAW_PREFIX)) {
    return `${API_BASE}${stored}`;
  }
  return value;
}

/** Point presigned PUT URLs at the same host the dashboard uses for API calls. */
export function rewriteUploadUrl(uploadUrl: string): string {
  try {
    const target = new URL(API_BASE);
    const parsed = new URL(uploadUrl);
    parsed.protocol = target.protocol;
    parsed.host = target.host;
    return parsed.toString();
  } catch {
    return uploadUrl;
  }
}

File: senam_app/lib/core/config/media_url.dart
import 'app_config.dart';

const _uploadsRawPrefix = '/v1/uploads/raw/';

const _uploadPurposes = [
  'profile_photo',
  'kyc_document',
  'review_photo',
  'company_logo',
  'portfolio_photo',
  'banner_image',
  'category_icon',
  'service_icon',
  'service_image',
];

bool _isUploadObjectKey(String value) {
  return _uploadPurposes.any((p) => value.startsWith('$p/'));
}

String toStoredMediaPath(String value) {
  final trimmed = value.trim();
  if (trimmed.isEmpty) return trimmed;

  final uri = Uri.tryParse(trimmed);
  if (uri != null && uri.hasScheme) {
    final idx = uri.path.indexOf(_uploadsRawPrefix);
    if (idx >= 0) return uri.path.substring(idx);
    return trimmed;
  }

  if (trimmed.startsWith(_uploadsRawPrefix)) return trimmed;

  if (_isUploadObjectKey(trimmed)) {
    return '$_uploadsRawPrefix$trimmed';
  }

  return trimmed;
}

String resolveMediaUrl(String? value) {
  if (value == null || value.isEmpty) return '';
  final base = AppConfig.apiBaseUrl.replaceAll(RegExp(r'/+$'), '');
  final stored = toStoredMediaPath(value);
  if (stored.startsWith(_uploadsRawPrefix)) {
    return '$base$stored';
  }
  return value;
}
